src/java.base/share/classes/java/util/concurrent/ConcurrentHashMap.java

```diff
@@ -1084,7 +1084,9 @@ else if (f instanceof ReservationNode)
      * @param m mappings to be stored in this map
      */
     public void putAll(Map<? extends K, ? extends V> m) {
-        tryPresize(m.size());
+        if (table != null) {
+            tryPresize(m.size());
```

Comment:
Shouldn't this be something like `tryPresize(this.size() + m.size())` to accommodate for the worst case where there are no common keys in `this` map and `m`?

Or do we intentionally try to be  conservative with memory usage? But then shouldn't we use at least `max(this.size(), m.size)`?

Reply:
> We don't need to compute max() here. [tryPresize()](https://github.com/openjdk/jdk/blob/8a4dc79e1a40e7115e2971af81623b6b0368f41c/src/java.base/share/classes/java/util/concurrent/ConcurrentHashMap.java#L2397) does that already.

The code you reference is only executed if `table == null` but in this case we know it is not `null` because we only call `tryPresize()` if `table != null`.

In general, it's hard fro me to understand what `tryPresize(int size)` is doing and if its `size` argument is supposed to contain the additional number of elements which will be inserted into the hash map or if it is a hint for the new total size of the hash map. Can you explain?

Reply:
> tryPresize(int size) is doing and if its size argument is supposed to contain the additional number of elements which will be inserted into the hash map or if it is a hint for the new total size of the hash map

Argument `size` for `tryPresize()` is a hint for the **new total size** of the hash map. If the size is too small compared to the current size of the map, there will be no resizing.

Reply:
Thanks. In that case, calling `tryPresize()` with `size < this.size()` doesn't make sense and we should call it with `this.size() + m.size()`.

Reply:
Thanks. Just committed this change. Also added a new benchmark for `putAll()` and the results in the top post.